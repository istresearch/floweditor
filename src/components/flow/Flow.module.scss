@import 'variables.module.scss';

.empty_flow {
  margin: 30px;
  width: 400px;

  h1 {
    font-family: 'Roboto', sans-serif;
    letter-spacing: 1px;
    padding: 0;
    text-transform: none;
    font-weight: 100;
    font-size: 18px;
    margin: 0;
  }

  div {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
  }
}