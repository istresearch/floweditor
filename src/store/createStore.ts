import { applyMiddleware, createStore, Middleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';

import rootReducer from 'store/rootReducer';
import AppState, { initialState } from 'store/state';

const middlewares: Middleware[] = [thunk];

export default (state: AppState = initialState) => {
  const store = createStore(
    rootReducer,
    state,
    composeWithDevTools(applyMiddleware(...middlewares))
  );

  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('./rootReducer', () => {
      const { default: nextRootReducer } = require('./rootReducer');
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
};
