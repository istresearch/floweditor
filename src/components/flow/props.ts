import { Type } from 'config/interfaces';
import { AnyAction, ContactProperties, FlowIssue } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { titleCase } from 'utils';

export interface IssueProps {
  helpArticles: { [key: string]: string };
  issues: FlowIssue[];
}

export interface ActionFormProps extends IssueProps {
  // action details
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;
  assetStore: AssetStore;

  addAsset(assetType: string, asset: Asset): void;

  // update handlers
  updateAction(
    action: AnyAction,
    onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
  ): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface RouterFormProps extends IssueProps {
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;

  assetStore: AssetStore;

  // update handlers
  updateRouter(renderNode: RenderNode): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface LocalizationFormProps extends IssueProps {
  language: Asset;
  nodeSettings: NodeEditorSettings;
  updateLocalizations(languageCode: string, localizations: any[]): void;
  onClose(canceled: boolean): void;
  helpArticles: { [key: string]: string };
}

export const NAME_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Name),
  id: ContactProperties.Name,
  type: AssetType.ContactProperty
};

export const CHANNEL_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Channel),
  id: ContactProperties.Channel,
  type: AssetType.ContactProperty
};

export const LANGUAGE_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Language),
  id: ContactProperties.Language,
  type: AssetType.ContactProperty
};

export const STATUS_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Status),
  id: ContactProperties.Status,
  type: AssetType.ContactProperty
};
