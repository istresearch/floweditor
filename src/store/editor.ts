import { FlowPosition, Contact } from 'flowTypes';
import ActionTypes, { UpdateEditorState } from 'store/actionTypes';
import Constants from 'store/constants';
import { Asset, RenderNode } from 'store/flowContext';

// tslint:disable:no-shadowed-variable
export interface DragSelection {
  startX?: number;
  startY?: number;
  currentX?: number;
  currentY?: number;
}

export interface DebugState {
  showUUIDs: boolean;
}

export interface CanvasPositions {
  [uuid: string]: FlowPosition;
}

export interface Activity {
  nodes: { [uuid: string]: number };
  segments: { [exitToNodeKey: string]: number };
  recentContacts?: { [key: string]: RecentContact[] };
  is_starting?: boolean;
}

export interface RecentContact {
  contact: Contact;
  operand: string;
  time: string;
}

export interface Warning {
  name: string;
  type: string;
}

export type Warnings = { [uuid: string]: Warning };

export interface EditorState {
  currentRevision: number | null;
  simulating: boolean;
  language: Asset | null;
  translating: boolean;
  fetchingFlow: boolean;
  ghostNode: RenderNode | null;
  dragActive: boolean;
  dragStartTime: number;
  dragDownPosition: FlowPosition | null;
  dragNodeUUID: string | null;
  dragGroup: boolean;
  dragSelection: DragSelection | null;
  debug?: DebugState | null;

  // which poptab is popped
  popped: string;

  warnings: Warnings;

  modalMessage?: ModalMessage;
  saving?: boolean;

  // the currently shown activity, can be
  // simulation or live
  activity: Activity;

  // the current live activity
  liveActivity: Activity;

  // interval in millis we should refresh activity
  activityInterval: number;

  // is our page visible or tabbed away
  visible: boolean;

  scrollToNode: string;
  scrollToAction: string;
}

export interface ModalMessage {
  title: string;
  body: string;
}

export const EMPTY_DRAG_STATE: any = {
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null
};

// Initial state
export const initialState: EditorState = {
  currentRevision: null,
  simulating: false,
  translating: false,
  language: null,
  fetchingFlow: false,
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null,
  ghostNode: null,
  debug: null,
  warnings: {},

  popped: null,

  activity: { segments: {}, nodes: {} },
  liveActivity: { segments: {}, nodes: {} },
  activityInterval: 5000,
  visible: true,

  scrollToNode: null,
  scrollToAction: null
};

// Action Creator
export const updateEditorState = (editorState: EditorState): UpdateEditorState => ({
  type: Constants.UPDATE_EDITOR_STATE,
  payload: {
    editorState
  }
});

/* Reducer */
export const editorState = (editorState: EditorState = initialState, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_EDITOR_STATE:
      return action.payload!.editorState;
    default:
      return editorState;
  }
};

// export our reducer
export default editorState;
