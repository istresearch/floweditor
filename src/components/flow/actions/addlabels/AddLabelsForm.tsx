import { default as bindCallbacks } from 'auto-bind/react';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { ActionFormProps } from 'components/flow/props';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import TypeList from 'components/nodeeditor/TypeList';
import { fakePropType } from 'config/ConfigProvider';
import * as React from 'react';
import { Asset } from 'store/flowContext';
import { FormState, mergeForm } from 'store/nodeEditor';
import { shouldRequireIf, validate } from 'store/validators';

import { initializeForm, stateToAction } from './helpers';
import i18n from 'config/i18n';
import { Trans } from 'react-i18next';
import { renderIssues } from '../helpers';

export interface AddLabelsFormState extends FormState {
  labels: any;
}

export const controlLabelSpecId = 'label';

export default class AddLabelsForm extends React.PureComponent<
  ActionFormProps,
  AddLabelsFormState
> {
  public static contextTypes = {
    assetService: fakePropType
  };

  constructor(props: ActionFormProps) {
    super(props);

    this.state = initializeForm(this.props.nodeSettings);
    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  public handleSave(): void {
    const valid = this.handleLabelsChanged(this.state.labels.value!, true);

    if (valid) {
      const newAction = stateToAction(this.props.nodeSettings, this.state);
      this.props.updateAction(newAction);
      this.props.onClose(false);
    }
  }

  public handleLabelsChanged(selected: Asset[], submitting: boolean = false): boolean {
    const updates: Partial<AddLabelsFormState> = {
      labels: validate(i18n.t('forms.labels', 'Labels'), selected, [shouldRequireIf(submitting)])
    };

    const updated = mergeForm(this.state, updates);
    this.setState(updated);
    return updated.valid;
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  public handleCreateAssetFromInput(input: string): any {
    return { name: input };
  }

  public handleLabelCreated(label: Asset): void {
    // update our store with our new group
    this.props.addAsset('labels', label);

    this.handleLabelsChanged(this.state.labels.value!.concat(label));
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;
    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p data-spec={controlLabelSpecId}>
          <Trans i18nKey="forms.add_labels_summary">
            Select the labels to apply to the incoming message.
          </Trans>
        </p>

        <AssetSelector
          name={i18n.t('forms.labels', 'Labels')}
          placeholder={i18n.t(
            'enter_to_create_label',
            'Enter the name of an existing label or create a new one'
          )}
          assets={this.props.assetStore.labels}
          entry={this.state.labels}
          searchable={true}
          multi={true}
          expressions={true}
          onChange={this.handleLabelsChanged}
          createPrefix={i18n.t('create_label', 'Create Label') + ': '}
          createAssetFromInput={this.handleCreateAssetFromInput}
          onAssetCreated={this.handleLabelCreated}
        />
        {renderIssues(this.props)}
      </Dialog>
    );
  }
}
