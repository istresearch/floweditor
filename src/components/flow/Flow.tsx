import { default as bindCallbacks } from 'auto-bind/react';
import Button from 'components/button/Button';
import { Canvas } from 'components/canvas/Canvas';
import { CanvasDraggableProps } from 'components/canvas/CanvasDraggable';
import Node from 'components/flow/node/Node';
import { getDraggedFrom } from 'components/helpers';
import NodeEditor from 'components/nodeeditor/NodeEditor';
import Simulator from 'components/simulator/Simulator';
import Sticky, { STICKY_BODY, STICKY_TITLE } from 'components/sticky/Sticky';
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { FlowDefinition, FlowMetadata, FlowPosition } from 'flowTypes';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Plumber from 'services/Plumber';
import { DragSelection, DebugState } from 'store/editor';
import { RenderNode } from 'store/flowContext';
import { createEmptyNode, detectLoops, getOrderedNodes } from 'store/helpers';
import { NodeEditorSettings } from 'store/nodeEditor';
import AppState from 'store/state';
import {
  ConnectionEvent,
  DispatchWithState,
  mergeEditorState,
  MergeEditorState,
  NoParamsAC,
  onConnectionDrag,
  OnConnectionDrag,
  OnOpenNodeEditor,
  onOpenNodeEditor,
  onRemoveNodes,
  OnRemoveNodes,
  OnUpdateCanvasPositions,
  onUpdateCanvasPositions,
  resetNodeEditingState,
  UpdateConnection,
  updateConnection,
  updateSticky,
  UpdateSticky
} from 'store/thunks';
import {
  createUUID,
  isRealValue,
  NODE_PADDING,
  renderIf,
  snapToGrid,
  timeEnd,
  timeStart
} from 'utils';
import Debug from 'utils/debug';

import styles from './Flow.module.scss';
import { Trans } from 'react-i18next';
import { PopTabType } from 'config/interfaces';
import i18n from 'config/i18n';

declare global {
  interface Window {
    fe: any;
  }
}

export interface FlowStoreProps {
  ghostNode: RenderNode;
  debug: DebugState;
  translating: boolean;
  popped: string;
  dragActive: boolean;

  mergeEditorState: MergeEditorState;

  definition: FlowDefinition;
  nodes: { [uuid: string]: RenderNode };
  metadata: FlowMetadata;
  nodeEditorSettings: NodeEditorSettings;

  updateConnection: UpdateConnection;
  onOpenNodeEditor: OnOpenNodeEditor;
  onUpdateCanvasPositions: OnUpdateCanvasPositions;
  onRemoveNodes: OnRemoveNodes;
  resetNodeEditingState: NoParamsAC;
  onConnectionDrag: OnConnectionDrag;
  updateSticky: UpdateSticky;
}

export interface Translations {
  [uuid: string]: any;
}

export const DRAG_THRESHOLD = 3;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
  return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
  const left = Math.min(drag.startX, drag.currentX);
  const top = Math.min(drag.startY, drag.currentY);
  const width = Math.max(drag.startX, drag.currentX) - left;
  const height = Math.max(drag.startY, drag.currentY) - top;
  return {
    left,
    top,
    width,
    height
  };
};

export class Flow extends React.PureComponent<FlowStoreProps, {}> {
  private Plumber: Plumber;
  private nodeContainerUUID: string;

  // Refs
  private ghost: any;

  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: FlowStoreProps, context: ConfigProviderContext) {
    super(props, context);

    this.nodeContainerUUID = createUUID();

    this.Plumber = new Plumber();

    /* istanbul ignore next */
    if (context.config.debug) {
      window.fe = new Debug(props, this.props.debug);
    }

    bindCallbacks(this, {
      include: [/Ref$/, /^on/, /^is/, /^get/, /^handle/]
    });

    timeStart('Loaded Flow');
  }

  private ghostRef(ref: any): any {
    return (this.ghost = ref);
  }

  public componentDidMount(): void {
    this.Plumber.bind('connection', (event: ConnectionEvent) =>
      this.props.updateConnection(event.sourceId, event.targetId)
    );
    this.Plumber.bind('beforeDrag', (event: ConnectionEvent) => {
      this.beforeConnectionDrag(event);
    });

    this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
      this.props.onConnectionDrag(event, this.context.config.flowType);
    });

    this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
      this.onConnectorDrop(event)
    );
    this.Plumber.bind(
      'beforeStartDetach',
      (event: ConnectionEvent) => !this.props.translating && this.context.config.mutable
    );
    this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
    this.Plumber.bind('beforeDrop', (event: ConnectionEvent) => this.onBeforeConnectorDrop(event));
    this.Plumber.triggerLoaded(this.context.config.onLoad);

    timeEnd('Loaded Flow');
  }

  public componentWillUnmount(): void {
    this.Plumber.reset();
  }

  /**
   * Called right before a connector is dropped onto a new node
   */
  private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
    this.props.resetNodeEditingState();
    const fromNodeUUID = event.sourceId.split(':')[0];
    try {
      detectLoops(this.props.nodes, fromNodeUUID, event.targetId);
    } catch {
      return false;
    }
    return true;
  }

  /**
   * Called the moment a connector is done dragging, whether it is dropped on an
   * existing node or on to empty space.
   */
  private onConnectorDrop(event: ConnectionEvent): boolean {
    const ghostNode = this.props.ghostNode;
    // Don't show the node editor if we a dragging back to where we were
    if (isRealValue(ghostNode) && !isDraggingBack(event)) {
      // Wire up the drag from to our ghost node
      this.Plumber.recalculate(ghostNode.node.uuid);

      const dragPoint = getDraggedFrom(ghostNode);

      this.Plumber.connect(dragPoint.nodeUUID + ':' + dragPoint.exitUUID, ghostNode.node.uuid);

      // Save our position for later
      const { left, top } = (this.ghost &&
        snapToGrid(this.ghost.ele.offsetLeft, this.ghost.ele.offsetTop)) || { left: 0, top: 0 };

      this.props.ghostNode.ui.position = { left, top };

      let originalAction = null;
      if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
        originalAction = ghostNode.node.actions[0];
      }

      // Bring up the node editor
      this.props.onOpenNodeEditor({
        originalNode: ghostNode,
        originalAction
      });
    }

    if (isDraggingBack(event)) {
      this.props.mergeEditorState({ ghostNode: null });
    }

    /* istanbul ignore next */
    document.removeEventListener('mousemove', (window as any).ghostListener);

    return true;
  }

  private beforeConnectionDrag(event: ConnectionEvent): boolean {
    if (event.source) {
      event.source.dispatchEvent(new Event('disconnect'));
    }
    return !this.props.translating;
  }

  private handleStickyCreation(props: CanvasDraggableProps) {
    const stickyMap = this.props.definition._ui.stickies || {};
    const uuid = props.uuid;
    return <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} selected={props.selected} />;
  }

  private handleNodeCreation(props: CanvasDraggableProps) {
    const onlyNode = Object.keys(this.props.nodes).length === 1;
    return (
      <Node
        onlyNode={onlyNode}
        startingNode={props.idx === 0}
        selected={props.selected}
        key={props.uuid}
        data-spec={nodeSpecId}
        nodeUUID={props.uuid}
        plumberMakeTarget={this.Plumber.makeTarget}
        plumberRemove={this.Plumber.remove}
        plumberRecalculate={this.Plumber.recalculate}
        plumberMakeSource={this.Plumber.makeSource}
        plumberConnectExit={this.Plumber.connectExit}
        plumberUpdateClass={this.Plumber.updateClass}
      />
    );
  }

  private getNodes(): CanvasDraggableProps[] {
    return getOrderedNodes(this.props.nodes).map((renderNode: RenderNode, idx: number) => {
      return {
        uuid: renderNode.node.uuid,
        position: renderNode.ui.position,
        elementCreator: this.handleNodeCreation,
        config: renderNode,
        idx
      };
    });
  }

  private getStickies(): CanvasDraggableProps[] {
    const stickyMap = this.props.definition._ui.stickies || {};
    return Object.keys(stickyMap).map((uuid: string, idx: number) => {
      return {
        uuid,
        elementCreator: this.handleStickyCreation,
        position: stickyMap[uuid].position,
        idx
      };
    });
  }

  private getDragNode(): JSX.Element {
    return isRealValue(this.props.ghostNode) ? (
      <div
        data-spec={ghostNodeSpecId}
        key={this.props.ghostNode.node.uuid}
        style={{ position: 'absolute', display: 'block', visibility: 'hidden' }}
      >
        <Node
          onlyNode={false}
          selected={false}
          startingNode={false}
          ref={this.ghostRef}
          ghost={true}
          nodeUUID={this.props.ghostNode.node.uuid}
          plumberMakeTarget={this.Plumber.makeTarget}
          plumberRemove={this.Plumber.remove}
          plumberRecalculate={this.Plumber.recalculate}
          plumberMakeSource={this.Plumber.makeSource}
          plumberConnectExit={this.Plumber.connectExit}
          plumberUpdateClass={this.Plumber.updateClass}
        />
      </div>
    ) : null;
  }

  private getSimulator(): JSX.Element {
    return renderIf(this.context.config.endpoints && this.context.config.endpoints.simulateStart)(
      <Simulator
        key="simulator"
        popped={this.props.popped}
        mergeEditorState={this.props.mergeEditorState}
        onToggled={(visible: boolean, tab: PopTabType) => {
          this.props.mergeEditorState({
            popped: visible ? tab : null
          });
        }}
      />
    );
  }

  private getNodeEditor(): JSX.Element {
    return renderIf(this.props.nodeEditorSettings !== null)(
      <NodeEditor
        key="node-editor"
        helpArticles={this.context.config.help}
        plumberConnectExit={this.Plumber.connectExit}
      />
    );
  }

  // TODO: this should be a callback from the canvas
  private handleDoubleClick(position: FlowPosition): void {
    const { left, top } = position;
    this.props.updateSticky(createUUID(), {
      position: snapToGrid(left - 90 + NODE_PADDING, top - 40),
      title: STICKY_TITLE,
      body: STICKY_BODY
    });
  }

  private getEmptyFlow(): JSX.Element {
    return (
      <div key="create_node" className={styles.empty_flow}>
        <Trans i18nKey="empty_flow_message">
          <h1>Let's get started</h1>
          <div>
            We recommend starting your flow by sending a message. This message will be sent to
            anybody right after they join the flow. This is your chance to send a single message or
            ask them a question.
          </div>
        </Trans>

        <Button
          name={i18n.t('buttons.create_message', 'Create Message')}
          onClick={() => {
            const emptyNode = createEmptyNode(null, null, 1, this.context.config.flowType);
            this.props.onOpenNodeEditor({
              originalNode: emptyNode,
              originalAction: emptyNode.node.actions[0]
            });
          }}
        />
      </div>
    );
  }

  /* 
  public componentDidUpdate(prevProps: FlowStoreProps): void {
    traceUpdate(this, prevProps);
  }
  */

  public handleDragging(uuids: string[]): void {
    uuids.forEach((uuid: string) => {
      try {
        const ele = document.getElementById(uuid);
        const exits = ele.querySelectorAll('.jtk-connected');
        this.Plumber.revalidate([ele, ...exits]);
      } catch (error) {}
    });
  }

  public handleCanvasLoaded(): void {
    this.Plumber.setContainer('canvas');
  }

  public render(): JSX.Element {
    const nodes = this.getNodes();

    const draggables = this.getStickies().concat(nodes);

    return (
      <div>
        {nodes.length === 0 ? this.getEmptyFlow() : <>{this.getSimulator()}</>}
        {this.getNodeEditor()}

        <Canvas
          mutable={this.context.config.mutable}
          draggingNew={!!this.props.ghostNode && !this.props.nodeEditorSettings}
          newDragElement={this.getDragNode()}
          onDragging={this.handleDragging}
          uuid={this.nodeContainerUUID}
          dragActive={this.props.dragActive}
          mergeEditorState={this.props.mergeEditorState}
          onRemoveNodes={this.props.onRemoveNodes}
          draggables={draggables}
          onDoubleClick={this.handleDoubleClick}
          onUpdatePositions={this.props.onUpdateCanvasPositions}
          onLoaded={this.handleCanvasLoaded}
        ></Canvas>
        <div id="activity_recent_contacts"></div>
      </div>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = ({
  flowContext: { definition, metadata, nodes },
  editorState: { ghostNode, debug, translating, popped, dragActive },
  // tslint:disable-next-line: no-shadowed-variable
  nodeEditor: { settings }
}: AppState) => {
  return {
    nodeEditorSettings: settings,
    definition,
    nodes,
    metadata,
    ghostNode,
    debug,
    translating,
    popped,
    dragActive
  };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
  bindActionCreators(
    {
      mergeEditorState,
      resetNodeEditingState,
      onConnectionDrag,
      onOpenNodeEditor,
      onUpdateCanvasPositions,
      onRemoveNodes,
      updateConnection,
      updateSticky
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Flow);
