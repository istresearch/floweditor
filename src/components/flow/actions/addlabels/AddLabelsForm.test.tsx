import { ActionFormProps } from 'components/flow/props';
import React from 'react';
import { AssetType, RenderNode } from 'store/flowContext';
import { fireEvent, render, fireChangeText, fireTembaSelect, getCallParams } from 'test/utils';
import { composeComponentTestUtils, mock } from 'testUtils';
import {
  createAddLabelsAction,
  createStartSessionAction,
  getActionFormProps,
  SubscribersGroup
} from 'testUtils/assetCreators';
import * as utils from 'utils';
import AddLabelsForm from './AddLabelsForm';

mock(utils, 'createUUID', utils.seededUUIDs());
const props = getActionFormProps(
  createAddLabelsAction([{ name: 'My Label', uuid: utils.createUUID() }])
);

describe(AddLabelsForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { baseElement, queryByTestId } = render(<AddLabelsForm {...props} />);
      expect(baseElement).toMatchSnapshot();
    });

    it('allows expressions', () => {
      const props = getActionFormProps(createStartSessionAction());
      const { baseElement, getByText, getByTestId } = render(<AddLabelsForm {...props} />);
      fireTembaSelect(getByTestId('temba_select_labels'), [
        {
          name: '@result.my_expression',
          expression: true
        }
      ]);

      expect(baseElement).toMatchSnapshot();

      // save our action
      fireEvent.click(getByText('Ok'));
      expect(props.updateAction).toHaveBeenCalled();
      expect(props.updateAction).toMatchCallSnapshot();

      const [addLabelAction] = getCallParams(props.updateAction);
      expect(JSON.stringify(addLabelAction)).toMatch('name_match');
    });
  });
});
