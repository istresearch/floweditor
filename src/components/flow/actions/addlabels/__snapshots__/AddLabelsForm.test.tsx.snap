// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLabelsForm render allows expressions 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header start_session"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Start Somebody Else
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"start_session\\",\\"name\\":\\"Start Somebody Else\\",\\"description\\":\\"Start somebody else in a flow\\",\\"localizeableKeys\\":[\\"exits\\"],\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"]}]"
              />
            </div>
          </div>
        </div>
        <p
          data-spec="label"
        >
          Select the labels to apply to the incoming message.
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              cachekey="0"
              data-testid="temba_select_labels"
              errors="[]"
              expressions="session"
              multi="true"
              name="Labels"
              namekey="name"
              placeholder="Enter the name of an existing label or create a new one"
              searchable="true"
              valuekey="uuid"
              values="[{\\"name\\":\\"@result.my_expression\\",\\"expression\\":true}]"
            />
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`AddLabelsForm render allows expressions 2`] = `
Array [
  Object {
    "labels": Array [
      Object {
        "name_match": "@result.my_expression",
      },
    ],
    "type": "add_input_labels",
    "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
  },
]
`;

exports[`AddLabelsForm render should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header add_input_labels"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Add Labels
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"add_input_labels\\",\\"name\\":\\"Add Labels\\",\\"description\\":\\"Label the incoming message\\",\\"visibility\\":[\\"messaging\\",\\"messaging_offline\\",\\"voice\\"]}]"
              />
            </div>
          </div>
        </div>
        <p
          data-spec="label"
        >
          Select the labels to apply to the incoming message.
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              cachekey="0"
              data-testid="temba_select_labels"
              errors="[]"
              expressions="session"
              multi="true"
              name="Labels"
              namekey="name"
              placeholder="Enter the name of an existing label or create a new one"
              searchable="true"
              valuekey="uuid"
              values="[{\\"name\\":\\"My Label\\",\\"uuid\\":\\"b1f332f3-bdd3-4891-aec5-1843a712dbf1\\"}]"
            />
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
